File: BOJ/1644.cpp
#include<iostream>
#include<vector>
using namespace std;

bool isPrime[4000001];
vector<int> primeNum;

void Eratos(int num) {
	for (int i = 0; i < 4000001; i++) {
		isPrime[i] = true;
	}

	for (int i = 2; i*i <= num; i++) {
		if (isPrime[i]) {
			for (int j = i*i; j <= num; j += i) {
				isPrime[j] = false;
			}
		}
	}

	for (int i = 2; i <= num; i++) {
		if (isPrime[i]) primeNum.push_back(i);
	}
}

int main() {
	int N;
	int answer = 0;
	cin >> N;

	Eratos(N);

	int left = 0, right = 0, sum = 0;
	while (left <= right) {
		if (sum >= N) sum -= primeNum[left++];
		else if (right == primeNum.size()) break;
		else sum += primeNum[right++];

		if (sum == N) answer++;
	}
	cout << answer << '\n';

	return 0;
}

File: BOJ/python/1644.py
import sys

input = sys.stdin.readline


def eratos(n):
    """
    Sieve of Eratosthenes

    Args:
        n (int): search prime number from 2 to n+1

    Returns:
        list: list of prime number
    """
    is_prime = [True] * (n + 1)
    for i in range(2, n + 1):
        if is_prime[i]:
            for j in range(i + i, n + 1, i):
                is_prime[j] = False

    return [i for i in range(2, n + 1) if is_prime[i]]


answer = 0
N = int(input().rstrip("\n"))


prime_list = eratos(N)
left = right = Sum = 0

# all of cases (prime number sum)
while left <= right:
    if Sum >= N:
        Sum -= prime_list[left]
        left += 1
    elif right == len(prime_list): # out of range
        break
    else:
        Sum += prime_list[right]
        right += 1

    if Sum == N:
        answer += 1

print(answer)
